# LeetCode Scramble

За задачку садился в течении нескольких вечеров по 2ч+-. Кроме контр примеров и некоторых их свойст, ничего интересного не установил. Принял решение сделать не оптимальный алгоритм решающий через поиск по всем возможным вариантам.

1. Строю древо всех возможных по условию задачи комбинаций.
2. Из древа получаю множество уникальных комбинаций.
3. Ищу по множеству требуемую комбинацию.

Думаю решение будет очень не оптимальное зато короткое. Далее буду оптимизировать и\или искать другое решение.

## Решение 1

In [9]:
import itertools as it
import math as ma

def mtl(s):
    ts = set([s,])
    if(len(s) == 1):
        return ts
    else:
        for i in range(len(s) - 1):
            Sa = mtl(s[0:i+1])
            Sb = mtl(s[i+1:len(s)])
            ts.update({*(''.join(t) for t in it.product(Sa, Sb)), *(''.join(t) for t in it.product(Sb, Sa))})
        return ts

In [2]:
t = mtl('abcd')
print(len(t), t)
print('cadb' in t)
print('bdac' in t)

22 {'dacb', 'bcda', 'cdab', 'cbad', 'abcd', 'acbd', 'adbc', 'dabc', 'bcad', 'dcba', 'dbca', 'dbac', 'acdb', 'cbda', 'bdca', 'abdc', 'badc', 'dcab', 'adcb', 'cabd', 'cdba', 'bacd'}
False
False


22 комбинации из 24 возможных, отсутсвует 'cadb' и зеркальное 'bdac'. На данном этапе понятно, что можно представить комбинациии букв как комбинации их индексов в исходной строке, тогда недоступные комбинации это комбинации вида - численный интервал (a,d) из N+<|S| переставленный в вид c,a,d,b. И если в комбинации присуствует любой такой интервал \*c\*a\*d\*b\* или зеркальный. Для букв встречающихся больше 1 раза, рассмотреть каждую возможную комбинацию индексов. Далее задачу буду решать в этом ключе.

Первое решение не прошло по time limit. На вот этих строчках.

In [4]:
#s1 = "abcdefghijklmn"
#s2 = "efghijklmncadb"
#len(s1)
#%time t = mtl(s1)
#s2 in t

CPU times: user 2min 50s, sys: 16.8 s, total: 3min 7s
Wall time: 3min 7s


У меня на компьютере занимает 3 минуты, сервер давал секунд 20. Видимо такая ветка решений вообще тупиковая по быстродействию.

## Решение 2

Строю алгоритм формирующий все тупиковые комбинации индексов из строки длинной n (3<n<31), пока оставляю все буквы уникальными. Потом буду решать вопрос с расстановкой индексов в строках s2 когда букве может соответствовать больше 1 индекса прим. 'cadbad', логика такая, если хотя бы одна расстановка индексов достижима то вся комбинация достижима.

In [30]:
#2031 - 3142
s1 = 'abcde'
s2 = 'cadbe'

t = mtl(s1)
print(f' \nВсего возможных комбинаций - {ma.factorial(len(s1))} \nКол-во достижимых комбинаций - {len(t)}  \n\nДостижимые комбинации - {t}')

In [31]:
t = mtl(s1)
print(f' \nВсего возможных комбинаций - {ma.factorial(len(s1))} \nКол-во достижимых комбинаций - {len(t)}  \n\nДостижимые комбинации - {t}')

 
Всего возможных комбинаций - 120 
Кол-во достижимых комбинаций - 90  

Достижимые комбинации - {'cdeab', 'bcade', 'edcba', 'acdeb', 'adcbe', 'cabde', 'acbed', 'ecdab', 'cdabe', 'adebc', 'bceda', 'baedc', 'deabc', 'edacb', 'cedab', 'dbcea', 'badec', 'abedc', 'abced', 'baced', 'dceba', 'ecabd', 'ecdba', 'aedcb', 'acdbe', 'ebcda', 'deacb', 'adecb', 'aecbd', 'dceab', 'edcab', 'dbace', 'cdbae', 'bedca', 'edbac', 'dacbe', 'decba', 'eacdb', 'dabce', 'aecdb', 'acedb', 'cabed', 'eabdc', 'cbdae', 'becda', 'ebadc', 'bacde', 'decab', 'eacbd', 'eadcb', 'aedbc', 'badce', 'debca', 'cdeba', 'bdcae', 'edabc', 'acbde', 'abecd', 'ebdca', 'bcaed', 'aebdc', 'dbcae', 'dcabe', 'cedba', 'bdeca', 'baecd', 'ebcad', 'cbdea', 'ecbad', 'abdce', 'cbeda', 'abdec', 'eabcd', 'abcde', 'edbca', 'bdcea', 'cbade', 'aebcd', 'dcbae', 'ebacd', 'eadbc', 'adceb', 'ecbda', 'bcdea', 'cbaed', 'cdbea', 'dcbea', 'bcdae', 'debac', 'adbce'}


In [27]:
s2 in t

False

На строках длинной 5 алгоритм должен дать 30 недостижимых комбинаций. 

In [43]:
s1 = 'abcdefgh'
s2 = 'cadbefgh'

t = mtl(s1)
fn = ma.factorial(len(s1))
an = len(t)

print(f' \nВсего возможных комбинаций - {fn} \nКол-во достижимых комбинаций - {an}  \n\nДостижимые комбинации - слшиком длинный вывод ')
print(f'Кол-во недостижимых комбинаций - {fn - an}')

 
Всего возможных комбинаций - 40320 
Кол-во достижимых комбинаций - 8558  

Достижимые комбинации - слшиком длинный вывод 
Кол-во недостижимых комбинаций - 31762


Кол-во недостижимых комбинаций, очень быстро растет с ростом n при n=5 u=30 при n = 8 u=32000, но мне нужны только те из них что длинной 4 их должно быть не много. Т.к. любая недостижимая комбинация большая 4, должна содержать в себе недостижимую комбинацию длинной 4.

Кол-во недостижимых комбинаций длинной 4 это кол-во интервалов длинной длинной 4 от 0 до n, где n длинна строки т.е n-4 и еще столько же зеркальных комбинаций (n-4)*2. Это нужно проверить на небольшом множестве.

In [84]:
s1 = 'abcde'
s2 = 'cadbe'

t = mtl(s1)
fn = ma.factorial(len(s1))
an = len(t)

print(f' \nВсего возможных комбинаций - {fn} \nКол-во достижимых комбинаций - {an}  \n\nДостижимые комбинации - слшиком длинный вывод ')
print(f'Кол-во недостижимых комбинаций - {fn - an}')

all_comb = {''.join(c) for c in it.permutations([*s1])}
unavailable_comb = all_comb.symmetric_difference(t)

 
Всего возможных комбинаций - 120 
Кол-во достижимых комбинаций - 90  

Достижимые комбинации - слшиком длинный вывод 
Кол-во недостижимых комбинаций - 30


In [91]:
#cadb - bdac
unavailable_comb

{'acebd',
 'adbec',
 'bcead',
 'bdace',
 'bdaec',
 'bdeac',
 'beacd',
 'beadc',
 'becad',
 'bedac',
 'cadbe',
 'cadeb',
 'caebd',
 'caedb',
 'cbead',
 'cdaeb',
 'ceabd',
 'ceadb',
 'cebad',
 'cebda',
 'dabec',
 'daceb',
 'daebc',
 'daecb',
 'dbaec',
 'dbeac',
 'dbeca',
 'dcaeb',
 'ebdac',
 'ecadb'}

Для строки длинной 5 получается 2а интервала a,b,c,d и b,c,d,e длинной 4 каждый интервал имеет зеркальный интервал. Всего 4 интервала которые можно спутать переставив символы как 
0,1,2,3 -> 2,0,3,1 из каждого недостижимого интервала можно получить комбинации добавляя все комбинации остальных свободных букв(индексов) между букв и с концов по схеме \*2\*3\*0\*1\*. Получается 20 комбинаций. Не понятно откуда еще 10. Надо проверять.

In [90]:
len(unavailable_comb)

30

Лишние 10 строк это новый вид недоступных комбинаций вида 1403(3041) индексы обладают теми же свойствами, что 2031, но не являются числами одного интервала, значит условие про итервал неверное. Принцип формирующий паттерны описанный ранее неполный или неверный. 

In [92]:
s1 = 'abcdef'
s2 = 'cadbef'

t = mtl(s1)
fn = ma.factorial(len(s1))
an = len(t)

print(f' \nВсего возможных комбинаций - {fn} \nКол-во достижимых комбинаций - {an}  \n\nДостижимые комбинации - слшиком длинный вывод ')
print(f'Кол-во недостижимых комбинаций - {fn - an}')

all_comb = {''.join(c) for c in it.permutations([*s1])}
unavailable_comb = all_comb.symmetric_difference(t)

 
Всего возможных комбинаций - 720 
Кол-во достижимых комбинаций - 394  

Достижимые комбинации - слшиком длинный вывод 
Кол-во недостижимых комбинаций - 326


При n = 6 кол-во недостижимых комбинаций из 30 становится 326. Считаю по старому паттерну, умножаю кол-во недоступных комб. длинной 4 на n+1 мест куда можно расставить остальные свободные буквы и умножаю на кол-во перестановок свободных букв.

In [99]:
n = 6
u_4_base = (n-3)*2
u_patterns_n = u_4_base*(n+1)*2
print(u_patterns_n)

84


Результат совершенно далекий от правды . .

Предполагаю, что общее у всех недоступных четверок при длинне строки (3<n<31) это соответсвие паттерну i2>i0>i3>i1, где i это индексы букв во входной строке S1. Прим. S1 = abcde по индексам s = 01234, пример недоступной строки согласно паттерну, сначало второй по величине индекс i0 = 3, затем самый маленький i1= 0, i2= 4, i3=1, получаем 3041, 2031 так же подходит. Новый паттерн лучше рассматриваю решения связанные с ним.